### ISSUE TITLE
Question: where does coc lie on the completion stack?

### ISSUE THREAD

**Issue by cprn** [neoclide/coc.nvim#1348]
**Is your feature request related to a problem? Please describe.**
No.

**Describe the solution you'd like**
I can submit a readme PR if somebody explains this to me and it's a frequent question.

**Describe alternatives you've considered**
(N/A)

**Additional context**
Sorry for asking this here but I'm having trouble understanding a basic principle of what `coc` is. Assuming `ale` + `deoplete` + `ultisnips` + `lsp` stack:
- does it replace `deoplete`?
- does it replace `ale`??
- can it work alongside `ale` (both support LSP)?
- can it work alongside `deoplete` (and is it recommended)?
- does it have its own templates??

I'm using `[dense-analysis/ale](https://github.com/dense-analysis/ale)` for last 2 months, it replaced `[autozimu/LanguageClient-neovim](https://github.com/autozimu/LanguageClient-neovim)` for me and I'm very happy with it. Now I heard about `coc`, spent some time reading the documentation and I'm still confused. There's an older issue trying to explain this but it predates `ale`'s LSP support. Also I've seen people calling `coc` a templating system (like UltiSnips, I presume?).

**Comment by fannheyward** [comment]
What do you mean _lie_?

1. `coc.nvim` is a LSP client that can talk to any language server. So coc can do all the functions that LS supports, includes code jump to definition, hover, completions suggest, lint, format, and so on.
2. But, not all languages server support all LSP features, so you can also use coc alongside ale lintting.
3. You can't use coc alongside deoplete
4. Besides LSP support, coc provides more features by extension, for example coc-emoji adds emoji completion source.

Checkout https://github.com/neoclide/coc.nvim/wiki for more.

**Comment by cprn** [comment]
*Lie* as in: where is it supposed to be placed. I think some of that is clear now. It replaces `deoplete` but has more features and can be enriched even more with extensions like `coc-snippets`, etc., which is very very cool, thank you!

However, I still need clarification on how it works alongside `ale`. When both are installed and I map `gd` to `<Plug>(ale_go_to_definition)` I get ALE talking directly to the Language Server. When I map `gd` to `<Plug>(coc-definition)` it's CoC talking directly to the Language Server. I don't understand what's the difference. Does CoC do something better I fail to notice?

If so, do I understand correctly that in order to benefit from both (ALE and CoC) it's recommended to use ALE without LSP support, map LSP related bindings to CoC's functions and set CoC to display diagnostics in ALE?

**Comment by fannheyward** [comment]
Personally, I suggest to use coc only if LS works.

**Comment by chemzqm** [comment]
> it's recommended to use ALE without LSP support, map LSP related bindings to coc's functions and set coc to display diagnostics in ALE?

You have to make the choice yourself, I don't know too much about ALE although I use it as linter.

**Comment by cprn** [comment]
This works surprisingly well! Sorry for being a bother. It's all clear now,
thank you very much for your help.

**Comment by horseinthesky** [comment]
I'm confused too by how to use CoC and ALE together.

Since I could not find out how to use CoC "real time" linting (can it do it?) which ALE is doing out of the box I decided to keep ALE for linting and CoC for formating:
```
{
  "coc.preferences.formatOnSaveFiletypes": [
    "*"
  ],
  "coc.preferences.extensionUpdateCheck": "never",
  "python.pythonPath": "/usr/bin/python3",
  "python.linting.enabled": false,
  "python.linting.pylintEnabled": false,
  "python.linting.flake8Enabled": true,
  "python.linting.flake8Args": [
    "--ignore=E501,E402,F401,E701"
  ],
  "python.formatting.provider": "autopep8",
  "python.formatting.autopep8Args": [
    "--ignore=E501"
  ],
  "yaml.validate": false,
  "yaml.format.enable": true,
  "yaml.completion": true,
  "prettier.formatterPriority": -1,
  "prettier.printWidth": 0
}
```

It works great for me but I just wonder if CoC could do this real time linting then I won't need ALE anymore.

**Comment by chemzqm** [comment]
coc.nvim support realtime lint, I don't know why it doesn't work for you.

**Comment by horseinthesky** [comment]
Could you help me discover what is wrong?

For instance with these coc-settings (linting is on):
```
{
  "coc.preferences.formatOnSaveFiletypes": [
    "*"
  ],
  "coc.preferences.extensionUpdateCheck": "never",
  "diagnostic.displayByAle": false,
  "python.pythonPath": "/usr/bin/python3",
  "python.linting.enabled": true,
  "python.linting.pylintEnabled": false,
  "python.linting.flake8Enabled": true,
  "python.linting.flake8Args": [
    "--ignore=E501,E402,F401,E701"
  ],
  "python.formatting.provider": "autopep8",
  "python.formatting.autopep8Args": [
    "--ignore=E501"
  ],
  "yaml.validate": true,
  "yaml.format.enable": true,
  "yaml.completion": true,
  "prettier.formatterPriority": -1,
  "prettier.printWidth": 0
}
```

And this block in init.vim:
```
" ==== coc ====
" Install CoC plugins
let g:coc_global_extensions = [
  \ 'coc-python',
  \ 'coc-snippets',
  \ 'coc-prettier',
  \ 'coc-json',
  \ 'coc-yaml',
  \ 'coc-css',
  \ 'coc-html',
  \ 'coc-xml',
  \ 'coc-tabnine',
  \ 'coc-go',
  \ ]

" use <tab> for trigger completion and navigate to the next complete item
function! s:check_back_space() abort
  let col = col('.') - 1
  return !col || getline('.')[col - 1]  =~ '\s'
endfunction

inoremap <silent><expr> <Tab>
  \ pumvisible() ? "\<C-n>" :
  \ <SID>check_back_space() ? "\<Tab>" :
  \ coc#refresh()

" Use <Tab> and <S-Tab> to navigate the completion list:
inoremap <expr> <Tab> pumvisible() ? "\<C-n>" : "\<Tab>"
inoremap <expr> <S-Tab> pumvisible() ? "\<C-p>" : "\<S-Tab>"
" Use <cr> to confirm completion
inoremap <expr> <cr> pumvisible() ? "\<C-y>" : "\<C-g>u\<CR>"
" To make <cr> select the first completion item and confirm the completion when no item has been selected:
" inoremap <silent><expr> <cr> pumvisible() ? coc#_select_confirm() : "\<C-g>u\<CR>"

" Navigate diagnostics
nmap <silent> [g <Plug>(coc-diagnostic-prev)
nmap <silent> ]g <Plug>(coc-diagnostic-next)
" Use <c-space>for trigger completion
inoremap <silent><expr> <c-space> coc#refresh()
" Show CoC extensions
nnoremap <silent> <space>e  :<C-u>CocList extensions<cr>
" Show commands
nnoremap <silent> <space>c  :<C-u>CocList commands<cr>

" Go to definition (Ctrl+6 for return)
nmap <silent> gd <Plug>(coc-definition)
" Format using linters
command! -nargs=0 Format :call CocAction('format')

" Add missing go imports on save
autocmd BufWritePre *.go :call CocAction('runCommand', 'editor.action.organizeImport')
" Add tags in Golang files
autocmd FileType go nmap ctj :CocCommand go.tags.add json<cr>
autocmd FileType go nmap cty :CocCommand go.tags.add yaml<cr>
autocmd FileType go nmap ctx :CocCommand go.tags.add xml<cr>
autocmd FileType go nmap ctc :CocCommand go.tags.clear<cr>
```

Linting shows something only after save, not in real time.

**Comment by fannheyward** [comment]
Which language doesn’t lint in real time? In my workflow, gopls can lint in real time.

**Comment by horseinthesky** [comment]
> Which language doesn’t lint in real time? In my workflow, gopls can lint in real time.
For me, it doesn't work for Python.

**Comment by cprn** [comment]
@horseinthesky 

If you can consistently find this bug on a fresh install of VIM and COC the best would be to submit a new issue and add the steps to reproduce. I can tell you right now python linting works for me with both ALE and COC (using coc as LSP client).

If you can't reproduce this on a fresh install it'd be better to ask on [gitter](https://gitter.im/neoclide/coc.nvim) - I'm personally guilty of submitting issues that could have been easily answered there without involving devs.

**Comment by horseinthesky** [comment]
I didn't know there is a place to discuss. Thanks.
I'll stick to ALE for know. I will check on a clean system later.
